[Bug] yarn add --peer after --dev doesn't add peer dependency

- [ ] I'd be willing to implement a fix

**Describe the bug**

I have a library that depends on `rxjs` as a peer dependency, but it also has tests inside it. 

If I add `rxjs` as a normal dependency, tests run properly. If instead I run:

`yarn add --peer --dev rxjs`

My tests will complain about: `Cannot find module 'rxjs' from '...'`

Note that I'm also using typescript and `ts-jest`. 

Ideally, the dependency should be available to both peer and dev dependencies.

**To Reproduce**

```js repro
const {promises: {readFile}} = require(`fs`);

await packageJsonAndInstall({
  devDependencies: {},
});

await yarn(`add`, `--dev`, `rxjs`);
await yarn(`add`, `--peer`, `rxjs`);

const pkgJson = JSON.parse(await readFile(`package.json`, `utf8`));
expect(pkgJson).toHaveProperty(`peerDependencies`);
expect(pkgJson).toHaveProperty(`devDependencies`);
expect(pkgJson.peerDependencies).toHaveProperty(`rxjs`);
expect(pkgJson.devDependencies).toHaveProperty(`rxjs`);
```

**Additional context**

This is also a bug in yarn v1: https://github.com/yarnpkg/yarn/issues/5287

--- a/.yarn/versions/de670c18.yml
+++ b/.yarn/versions/de670c18.yml
@@ -1,0 +1,3 @@
+releases:
+  "@yarnpkg/cli": prerelease
+  "@yarnpkg/plugin-essentials": prerelease

--- a/.yarn/versions/de670c18.yml
+++ b/.yarn/versions/de670c18.yml
@@ -1,3 +1,19 @@
 releases:
   "@yarnpkg/cli": prerelease
   "@yarnpkg/plugin-essentials": prerelease
+
+declined:
+  - "@yarnpkg/plugin-constraints"
+  - "@yarnpkg/plugin-dlx"
+  - "@yarnpkg/plugin-init"
+  - "@yarnpkg/plugin-interactive-tools"
+  - "@yarnpkg/plugin-npm-cli"
+  - "@yarnpkg/plugin-pack"
+  - "@yarnpkg/plugin-patch"
+  - "@yarnpkg/plugin-pnp"
+  - "@yarnpkg/plugin-stage"
+  - "@yarnpkg/plugin-typescript"
+  - "@yarnpkg/plugin-version"
+  - "@yarnpkg/plugin-workspace-tools"
+  - "@yarnpkg/check"
+  - "@yarnpkg/core"

--- a/packages/acceptance-tests/pkg-tests-specs/sources/commands/add.test.js
+++ b/packages/acceptance-tests/pkg-tests-specs/sources/commands/add.test.js
@@ -258,6 +258,23 @@ describe(`Commands`, () => {
     );
 
     test(
+      `it should add an existing development dependency as peer dependency when added using --peer `,
+      makeTemporaryEnv({}, async ({path, run, source}) => {
+        await run(`add`, `no-deps`, `-D`);
+        await run(`add`, `no-deps`, `-P`);
+
+        await expect(xfs.readJsonPromise(`${path}/package.json`)).resolves.toMatchObject({
+          devDependencies: {
+            [`no-deps`]: `^2.0.0`,
+          },
+          peerDependencies: {
+            [`no-deps`]: `*`,
+          },
+        });
+      }),
+    );
+
+    test(
       `it should add a node-gyp dependency to the lockfile if a script uses it`,
       makeTemporaryEnv({}, async ({path, run, source}) => {
         await run(`add`, `inject-node-gyp`);

--- a/packages/plugin-essentials/sources/commands/add.ts
+++ b/packages/plugin-essentials/sources/commands/add.ts
@@ -246,14 +246,14 @@ function suggestTarget(workspace: Workspace, ident: Ident, {dev, peer, preferDev
   if (!dev && !peer && hasPeer)
     throw new UsageError(`Package "${structUtils.prettyIdent(workspace.project.configuration, ident)}" is already listed as a peer dependency - use either of -D or -P, or remove it from your peer dependencies first`);
 
-  if (hasRegular)
-    return suggestUtils.Target.REGULAR;
-  if (hasDev)
-    return suggestUtils.Target.DEVELOPMENT;
-
   if (peer)
     return suggestUtils.Target.PEER;
   if (dev || preferDev)
+    return suggestUtils.Target.DEVELOPMENT;
+
+  if (hasRegular)
+    return suggestUtils.Target.REGULAR;
+  if (hasDev)
     return suggestUtils.Target.DEVELOPMENT;
 
   return suggestUtils.Target.REGULAR;

